provider/openstack/cinder.go

```diff
@@ -0,0 +1,443 @@
+
+// VolumeSource implements storage.Provider.
+func (p *OpenstackProvider) VolumeSource(environConfig *config.Config, providerConfig *storage.Config) (storage.VolumeSource, error) {
+
+	if err := p.ValidateConfig(providerConfig); err != nil {
+		return nil, err
+	}
+
+	openstackClient, err := p.newOpenstackAdapter(environConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &openstackVolumeSource{openstackClient}, nil
+}
+
+// FilesystemSource implements storage.Provider.
+func (p *OpenstackProvider) FilesystemSource(environConfig *config.Config, providerConfig *storage.Config) (storage.FilesystemSource, error) {
+	return nil, errors.NotSupportedf("filesystems")
+}
+
+// Supports implements storage.Provider.
+func (p *OpenstackProvider) Supports(kind storage.StorageKind) bool {
+	switch kind {
+	case storage.StorageKindBlock:
+		return true
+	}
+	return false
+}
+
+// Scope implements storage.Provider.
+func (s *OpenstackProvider) Scope() storage.Scope {
+	return storage.ScopeEnviron
+}
+
+// ValidateConfig implements storage.Provider.
+func (p *OpenstackProvider) ValidateConfig(cfg *storage.Config) error {
+	return nil
+}
+
+// Dynamic implements storage.Provider.
+func (p *OpenstackProvider) Dynamic() bool {
+	return true
+}
+
+type openstackVolumeSource struct {
+	openstackClient OpenstackAdapter
+}
+
+var _ storage.VolumeSource = (*openstackVolumeSource)(nil)
+
+// CreateVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) CreateVolumes(args []storage.VolumeParams) (vols []storage.Volume, volAtt []storage.VolumeAttachment, returnErr error) {
+
+	vols = make([]storage.Volume, 0, len(args))
+	volAtt = make([]storage.VolumeAttachment, 0, len(args))
+
+	for _, a := range args {
+
+		createdVol, err := s.openstackClient.CreateVolume(a.Size, a.Tag)
+		if err != nil {
+			return nil, nil, err
+		}
+		vols = append(vols, createdVol)
+
+		// If the method exits with an error, be sure to delete any
+		// created volumes so that we're idempotent. We create several
+		// clousures instead of one to take advantage of the loop
+		// parameter.
+		defer func(arg storage.VolumeParams, createdVol storage.Volume) {
+			if returnErr == nil {
+				return
+			}
+
+			if arg.Attachment != nil {
+				volAttachments, err := s.openstackClient.ListVolumeAttachments(string(arg.Attachment.InstanceId))
+				if err != nil {
+					logger.Warningf("could not list volumes while cleaning up: %v", err)
+					return
+				}
+
+				if err := detachVolume(
+					string(arg.Attachment.InstanceId),
+					arg.Attachment.VolumeId,
+					volAttachments,
+					s.openstackClient,
+				); err != nil {
+					logger.Warningf("could not detach volumes while cleaning up: %v", err)
+					return
+				}
+			}
+
+			if err := s.openstackClient.DeleteVolume(createdVol.VolumeId); err != nil {
+				logger.Warningf("could not delete volumes while cleaning up: %v", err)
+				return
+			}
+		}(a, createdVol)
+
+		if a.Attachment != nil {
```

Comment:
`&& a.Attachment.InstanceId != ""`

Because this is a dynamic provider, the instance may not be provisioned yet. This should change in the future (i.e. we should wait until the instance is provisioned before calling CreateVolumes with an attachment), but until then we need to check.